File: src/core/feedback/collector.ts
/**
 * 反馈收集器
 */

import { db } from '@/lib/db';

export async function addFeedback(params: {
    userId?: string;
    sessionId: string;
    messageId?: string;
    rating: number;
    comment?: string;
    query?: string;
    response?: string;
    intent?: string;
    domain?: string;
}): Promise<boolean> {
    try {
        await db.feedback.create({
            data: {
                userId: params.userId,
                sessionId: params.sessionId,
                messageId: params.messageId,
                rating: params.rating,
                comment: params.comment,
                query: params.query,
                response: params.response,
                intent: params.intent,
                domain: params.domain,
            },
        });
        return true;
    } catch (e) {
        console.error('[FeedbackCollector] 添加反馈失败:', e);
        return false;
    }
}

export async function getLowQualitySamples(domain?: string, limit: number = 20) {
    try {
        return await db.feedback.findMany({
            where: {
                rating: -1,
                ...(domain && { domain }),
            },
            orderBy: { createdAt: 'desc' },
            take: limit,
        });
    } catch (e) {
        console.error('[FeedbackCollector] 获取低质量样本失败:', e);
        return [];
    }
}

export async function getFeedbackStats(domain?: string) {
    try {
        const positive = await db.feedback.count({
            where: { rating: 1, ...(domain && { domain }) },
        });
        const negative = await db.feedback.count({
            where: { rating: -1, ...(domain && { domain }) },
        });

        return {
            positive,
            negative,
            total: positive + negative,
            satisfactionRate:
                positive + negative > 0
                    ? ((positive / (positive + negative)) * 100).toFixed(1) + '%'
                    : 'N/A',
        };
    } catch (e) {
        console.error('[FeedbackCollector] 获取统计失败:', e);
        return { positive: 0, negative: 0, total: 0, satisfactionRate: 'N/A' };
    }
}

File: server/core/feedback/collector.py
"""
Feedback collector module for Lingxi backend.
"""
from datetime import datetime
from typing import Dict, Any, List, Optional

_feedbacks: List[Dict[str, Any]] = []


async def add_feedback(feedback: Dict[str, Any]) -> bool:
    feedback["created_at"] = datetime.now().isoformat()
    _feedbacks.append(feedback)
    return True


async def get_feedback_stats(domain: Optional[str] = None) -> Dict[str, Any]:
    feedbacks = _feedbacks
    if domain:
        feedbacks = [f for f in feedbacks if f.get("domain") == domain]
    total = len(feedbacks)
    positive = sum(1 for f in feedbacks if f.get("rating") == 1)
    negative = sum(1 for f in feedbacks if f.get("rating") == -1)
    intent_counts: Dict[str, int] = {}
    for f in feedbacks:
        intent = f.get("intent", "unknown")
        intent_counts[intent] = intent_counts.get(intent, 0) + 1
    return {
        "total": total,
        "positive": positive,
        "negative": negative,
        "satisfaction_rate": round(positive / total * 100, 1) if total > 0 else 0,
        "intent_distribution": intent_counts,
    }


async def get_recent_feedback(limit: int = 50, domain: Optional[str] = None) -> List[Dict[str, Any]]:
    feedbacks = _feedbacks
    if domain:
        feedbacks = [f for f in feedbacks if f.get("domain") == domain]
    return feedbacks[-limit:]